src/dotty/tools/dotc/parsing/Parsers.scala

```diff
@@ -536,30 +536,38 @@ object Parsers {
      */
     def path(thisOK: Boolean, finish: Tree => Tree = id): Tree = {
       val start = in.offset
-      def handleThis(name: TypeName) = {
+      var tTpeName: TypeName = tpnme.EMPTY
+
+      def acceptThis(): Tree = {
         in.nextToken()
-        val t = atPos(start) { This(name) }
-        if (!thisOK && in.token != DOT) syntaxError("`.' expected")
-        manyDotSelectors(t, finish)
+        atPos(start) { This(tTpeName) }
       }
-      def handleSuper(name: TypeName) = {
+      def acceptSuper(): Tree = {
         in.nextToken()
         val mix = mixinQualifierOpt()
-        val t = atPos(start) { Super(This(name), mix) }
-        accept(DOT)
-        manyDotSelectors(selector(t), finish)
+        atPos(start) { Super(This(tTpeName), mix) }
       }
-      if (in.token == THIS) handleThis(tpnme.EMPTY)
-      else if (in.token == SUPER) handleSuper(tpnme.EMPTY)
-      else {
+      def thisSuperOrElse(alternative: => Tree): Tree =
+        if (in.token == THIS && !thisOK)
+          someDotSelectors(acceptThis(), finish)
+        else if (in.token == THIS)
+          manyDotSelectors(acceptThis(), finish)
```

Comment:
Yes, the distinction is tricky -- but I tried to make it explicit in the name. In contrast, before the "Kleene-plus" nature of `someDotSelectors` was inlined in the implementation of `path`.